---
title: "Shigella RNA-seq pilot experiment, edgeR"
author: "Laura Jenniches"
date: "24/5/2024"
output:
  pdf_document: default
  html_document: default
---

```{r global-options, include=FALSE}
# set result directory for plots etc.
res_dir <- "../results/pilot_RNAseq_edgeR/"
data_dir <- "../data/"
```

```{r setup, results="hide"}
## Bioconductor
library('edgeR')

## CRAN
library('ggplot2')
library('gplots') # heatmap.2
library('RColorBrewer')
library('pheatmap')
library('dplyr')
library('data.table')
library(extrafont)
# font_import("y")
# fonts()
loadfonts()

theme_set(theme_bw(base_size = 10, base_family="Arial"))

cb_palette <- c("#864C95", "#5876A3", "#9BD0BB", "#FFD966", "#E76570")
cb_palette2 <- c("#5876A3", "#F0845D", "gray", "darkred", "darkblue", "#9BD0BB")
cb_palette3 <- c("#424242", "#d36c6c")
```



```{r}
meta_data <- fread("../annotations/Shigella_meta_combined.csv", quote="\"", stringsAsFactors=F, data.table=F)
rownames(meta_data) <- meta_data$locus_tag
genes_chromosome <- meta_data$locus_tag[which(meta_data$seq == unique(meta_data$seq)[1])]
genes_plasmid <- meta_data$locus_tag[which(meta_data$seq == unique(meta_data$seq)[2])]

read_counts <- fread(paste0(data_dir, "counts_pilot_RNAseq.csv.gz"), quote="\"", stringsAsFactors=F, data.table=F)
rownames(read_counts) <- read_counts$locus_tag
read_counts <- read_counts[,-c(1)]

cond_data <- sapply(colnames(read_counts), function(x) strsplit(as.character(x), "[_]")[[1]])
cond <- data.frame(sample=sub("S","", cond_data[4,]),
                   strain=cond_data[2,],
                   replicate=as.integer(cond_data[1,]),
                   temp=cond_data[3,],
                   identifier=colnames(read_counts),
                   row.names=colnames(read_counts),
                   stringsAsFactors=F)

N_sa <- nrow(cond)
```

```{r}
group <- factor(paste(cond$strain, cond$temp, sep="_"))
dgeAll <- DGEList(read_counts %>% as.matrix(), group=group, genes=rownames(read_counts))
keep <- rowSums(edgeR::cpm(dgeAll$counts[,!grepl("RNA", colnames(dgeAll$counts))]) > 10) >= 3
dgeAll <- dgeAll[keep, , keep.lib.sizes=FALSE]

dgeAll <- calcNormFactors(dgeAll, method="TMM")
normCounts <- edgeR::cpm(dgeAll)
pseudoNormCounts <- log2(normCounts + 1)

design <- model.matrix(~0+group)
colnames(design) <- levels(group)
```

# Normalization
The data was normalized using edgeR trimmed mean of M-values (TMM).

```{r}
pseudoCounts <- log2(dgeAll$counts+1)
boxplot(pseudoCounts, col="gray", las=3, main="Before normalization")
```

```{r}
boxplot(pseudoNormCounts, col="gray", las=3, main="After TMM normalization")
```

# Clustering of samples
## Multidimensional scaling (MDS) plot

```{r, fig.show="hide"}
opts_chunk$set(fig.show="hide")
mds_data <- plotMDS(pseudoNormCounts)
```

```{r, fig.show="asis"}
mds_data <- data.frame(sample=dgeAll$samples %>% rownames(), group=group, dim1=mds_data$x, dim2=mds_data$y)
g <- ggplot(mds_data, aes(x=dim1, y=dim2)) +
#  geom_point(aes(col=group)) +
  geom_text(aes(label=sample, col=group)) +
  scale_color_discrete(guide=F) +
  theme(rect=element_rect(fill="transparent"),
        axis.line=element_line(linewidth=unit(0.1, "pt")),
        legend.position="top", legend.margin=margin(t=-0.2, b=-0.2, unit="cm"),
        legend.key.width=unit(0.3,"cm"),legend.key.height=unit(0.3,"cm"),
        strip.background = element_rect(fill=NA, size=0.5), panel.spacing = unit(0, "lines"),
        panel.grid=element_blank(), axis.text=element_text(color = "black"))
plot(g)
```

# Clustering of samples
We have already seen that the samples cluster nicely in the MDS plot. Another way to cluster the samples by sample distance which was calculated using normalized log2 counts. Similar samples have small distances (i.e. simililar normalized read counts). The results are presented in form of a heatmap. Also here, replicates nicely cluster together. In addition, we can see that the different treatment of the two experimental blocks leads to smaller differences within one experimental block, and that samples cultured at the same temperature cluster together. Already here, it becomes apparent that the gene expressions in the virF deletion mutant differ much more at 37C than at 30C, where virulence genes are not active.

```{r, fig.show="asis"}
sampleDists <- dist(t(pseudoNormCounts))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- colnames(pseudoCounts)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```


```{r, fig.show="asis", eval=F}
# use cpm values instead of read counts
scaledata <- t(scale(t(normCounts))) # Centers and scales data.
scaledata <- scaledata[complete.cases(scaledata),]

hr <- hclust(as.dist(1-cor(t(scaledata), method="pearson")), method="complete") # Cluster rows by Pearson correlation.
hc <- hclust(as.dist(1-cor(scaledata, method="spearman")), method="complete") # Clusters columns by Spearman correlation.

heatmap.2(normCounts,
          Rowv=as.dendrogram(hr), 
          Colv=as.dendrogram(hc),
          col=redgreen(100),
          scale="row",
          margins = c(7, 7),
          cexCol = 0.7,
          labRow = F,
          main = "First experimental block",
          trace = "none")
```

# Differential expression analysis: MA plots and top tags
The MA plots show the genes where read counts differ significantly at an FDR of 0.01. The directories which contain the MA plots and their corresponding tables are given below.

```{r, fig.show="asis"}
dgeAll <- estimateDisp(dgeAll, design, robust=TRUE)
# plotBCV(dgeAll)

fit <- glmQLFit(dgeAll,design,robust=T)
# plotQLDisp(fit)
con <- makeContrasts(M90TvsFd_37C = M90T_37C - Fd_37C, #1
                     M90TvsFd_30C = M90T_30C - Fd_30C, #2
                     M90T_37vs30C = M90T_37C - M90T_30C, #3
                     levels=design)
contrast_info <- c(bquote("1) WT vs. " ~ Delta * "virF at" ~ 37^degree * "C"), #1
                   bquote("2) WT vs. " ~ Delta * "virF at" ~ 30^degree * "C"), #2
                   bquote("3) WT at" ~ 37^degree * "C vs." ~ 30^degree * "C")) #3
names(contrast_info) <- colnames(con)
```

##First experimental block

First of all, we determine the transcripts which are differentially expressed in the $\Delta$virF mutant (Fd) strain compared to the WT (M90T) strain at 30 and 37C. In addition, we compare WT at 37C to 30C. The result of this can be compared to the TraDIS results for the M90T strain at different conditions. The corresponding comparisons are (in parenthesis, the name of the contrast which correspond to the names of the directories are given)

1) WT vs. $\Delta$virF at 37C (M90TvsFd_37C = M90T_37C - Fd_37C)
2) WT vs. $\Delta$virF at 30C (M90TvsFd_30C = M90T_30C - Fd_30C)
3) WT at 37C vs. 30C (M90T_37vs30C = M90T_37C - M90T_30C)

```{r, fig.show="asis"}
for (a_con in colnames(con)) {
  qlf <- glmQLFTest(fit, contrast=con[,a_con])
  # tr <- glmTreat(fit, contrast=con[,a_con], lfc=1)
  tt <- topTags(qlf, Inf) %>% as.data.frame()
  tt_mod <- cbind(locus_tag=rownames(tt),
                  gene_name=meta_data$com_id[match(rownames(tt), meta_data$locus_tag)],
                  description=meta_data$description[match(rownames(tt), meta_data$locus_tag)],
                  type=meta_data$type[match(rownames(tt), meta_data$locus_tag)],
                  tt[,c(2:3,5:6)],
                  m_log10FDR=-log10(tt$FDR))
  # all genes, ordered by log-fold change
  write.table(tt_mod, paste0(res_dir, "/tt_", a_con, ".csv"), row.names=F, quote=F, sep="\t")
  # gene list chromosome
  write.table(subset(tt_mod, locus_tag %in% genes_chromosome), paste0(res_dir, "/tt_chromosome_", a_con, ".csv"), row.names=F, quote=F, sep="\t")
  # gene list virulence plasmid
  write.table(subset(tt_mod, locus_tag %in% genes_plasmid), paste0(res_dir, "/tt_plasmid_", a_con, ".csv"), row.names=F, quote=F, sep="\t")
  
  pos.sig <- which(tt_mod$FDR < 0.01 & tt_mod$logFC>1)
  neg.sig <- which(tt_mod$FDR < 0.01 & tt_mod$logFC < -1) 
  
  p_MA <- tt_mod %>% ggplot(aes(logCPM, logFC)) +
    geom_point(col=cb_palette2[3], alpha=0.4) +
    geom_point(aes(col="adv. FDR\u22640.01"), shape=1, size=1.5, data=subset(tt_mod, FDR<=0.01 & logFC < -1)) +
    geom_point(aes(col="detr. FDR\u22640.01"), shape=1, size=1.5, data=subset(tt_mod, FDR<=0.01 & logFC>1)) +
    scale_color_manual(name="", values=cb_palette2[c(1,2)], breaks=c("adv. FDR\u22640.01", "detr. FDR\u22640.01")) +
    theme(rect=element_rect(fill="transparent"),
        axis.line=element_line(linewidth=unit(0.1, "pt")),
        legend.position="top", legend.margin=margin(t=-0.2, b=-0.2, unit="cm"),
        legend.key.width=unit(0.3,"cm"),legend.key.height=unit(0.3,"cm"),
        strip.background = element_blank(), panel.spacing = unit(0, "lines"),
        panel.grid=element_blank(), axis.text=element_text(color = "black"))
  print(p_MA)
  cairo_pdf(paste0(res_dir, "/MA_", a_con, ".pdf"), width = 3, height = 3)
  print(p_MA)
  dev.off()
  
  p_max <- tt_mod %>% subset(FDR<=0.01 & abs(logFC)>1) %>% pull(PValue) %>% max()
  p_v <- tt_mod %>% ggplot(aes(logFC, -log10(PValue))) +
    geom_point(col=cb_palette2[3], alpha=0.4) +
    geom_point(aes(col="adv. FDR\u22640.01"), shape=1, size=1.5, data=subset(tt_mod, FDR<=0.01 & logFC < -1)) +
    geom_point(aes(col="detr. FDR\u22640.01"), shape=1, size=1.5, data=subset(tt_mod, FDR<=0.01 & logFC>1)) +
    geom_hline(yintercept=-log10(p_max), linetype="dashed") +
    geom_vline(xintercept=-1, col=cb_palette2[1], linetype="dashed") +
    geom_vline(xintercept=1, col=cb_palette2[2], linetype="dashed") +
    scale_color_manual(name="", values=cb_palette2[c(1,2)], breaks=c("adv. FDR\u22640.01", "detr. FDR\u22640.01")) +
    theme(rect=element_rect(fill="transparent"),
        axis.line=element_line(linewidth=unit(0.1, "pt")),
        legend.position="top", legend.margin=margin(t=-0.2, b=-0.2, unit="cm"),
        legend.key.width=unit(0.3,"cm"),legend.key.height=unit(0.3,"cm"),
        strip.background = element_blank(), panel.spacing = unit(0, "lines"),
        panel.grid=element_blank(), axis.text=element_text(color = "black"))
  print(p_v)
  cairo_pdf(paste0(res_dir, "/volcano_", a_con, ".pdf"), width = 2, height = 2.1)
  print(p_v)
  dev.off()
}
```
